### Lab Python 5-1 | Customer Analysis Round 1

In [1]:
import pandas as pd
import numpy as np

In [2]:
#Read the three files into python as dataframes
df1=pd.read_csv("Data/file1.csv")
df2=pd.read_csv("Data/file2.csv")
df3=pd.read_csv("Data/file3.csv")

In [3]:
#Show the DataFrame's shape
print(df1.shape)
print(df2.shape)
print(df3.shape)

(4008, 11)
(996, 11)
(7070, 11)


In [4]:
df1.sample()

Customer   ST GENDER Education Customer Lifetime Value  Income  \
1596      NaN  NaN    NaN       NaN                     NaN     NaN   

      Monthly Premium Auto Number of Open Complaints Policy Type  \
1596                   NaN                       NaN         NaN   

     Vehicle Class  Total Claim Amount  
1596           NaN                 NaN

In [5]:
#Standardize header names
#Paso los títulos de las columnas a minúsculas y cambio los espacios por barras bajas a modo de respetar el PEP 8 de Python.
df1.columns=df1.columns.str.lower().str.replace(" ", "_")
df2.columns=df2.columns.str.lower().str.replace(" ", "_")
df3.columns=df3.columns.str.lower().str.replace(" ", "_")

In [6]:
df1.sample()

customer   st gender education customer_lifetime_value  income  \
2665      NaN  NaN    NaN       NaN                     NaN     NaN   

      monthly_premium_auto number_of_open_complaints policy_type  \
2665                   NaN                       NaN         NaN   

     vehicle_class  total_claim_amount  
2665           NaN                 NaN

In [7]:
df1.rename(columns={"st":"state"}, inplace=True)
df2.rename(columns={"st":"state"}, inplace=True)
# df3 ya lo tiene bien.

In [8]:
df1.sample()

customer state gender education customer_lifetime_value  income  \
3812      NaN   NaN    NaN       NaN                     NaN     NaN   

      monthly_premium_auto number_of_open_complaints policy_type  \
3812                   NaN                       NaN         NaN   

     vehicle_class  total_claim_amount  
3812           NaN                 NaN

In [9]:
df3.columns

Index(['customer', 'state', 'customer_lifetime_value', 'education', 'gender',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'total_claim_amount', 'vehicle_class'],
      dtype='object')

In [10]:
#Rearrange the columns in each dataframe as needed
#Dado que los 3 files tienen las mismas columnas pero en diferentes órdenes, elijo reordenar las columnas de df2 y df3 para que queden como en df1 así después se puede concatenar correctamente.
df2=df2[df1.columns]
df3=df3[df1.columns]

print(df1.columns)
print(df2.columns)
print(df3.columns)

Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')
Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')
Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')


In [11]:
#Expresión booleana para corroborar
booleana= df1.columns.equals(df2.columns) and df1.columns.equals(df3.columns)
booleana

True

Chequeemos si los shapes de los DataFrames luego de todos los cambios siguen siendo iguales a los calculados inicialmente:


In [12]:
df1.shape

(4008, 11)

In [13]:
df2.shape

(996, 11)

In [14]:
df3.shape

(7070, 11)

In [15]:
print("\u2713")

✓


In [16]:
# Concatenate the 3 DataFrames
los3=pd.concat([df1, df2, df3], axis=0) 
los3.shape

(12074, 11)

In [17]:
# Which columns are numerical?
los3.sample()

customer   state gender education customer_lifetime_value   income  \
4523  YG71375  Nevada      F   College             5964.345179  98078.0   

      monthly_premium_auto number_of_open_complaints    policy_type  \
4523                  73.0                         0  Personal Auto   

      vehicle_class  total_claim_amount  
4523  Four-Door Car          418.689402

In [18]:
los3.dtypes

customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value       object
income                       float64
monthly_premium_auto         float64
number_of_open_complaints     object
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object

Las que seguro son numerical son las que aparecen en float64 en el output de la celda de arriba (income, monthly_premium_auto y total_claim_amount).
Luego hay 2 que son dudosas..
customer_lifetime_value y number_of_open_complaints. Estrictamente no son numéricas ya que Python las interpreta como object. Sin embargo, los humanos sabemos que sí lo son (Humanos 1 - Programas+ AIs 0). Por ejemplo, la de customer_lifetime_values en algunos casos tiene "%"s, eso podría estar afectando.

In [19]:
#Por las dudas
por_las_dudas=los3._get_numeric_data().columns.tolist()
print(por_las_dudas)

['income', 'monthly_premium_auto', 'total_claim_amount']


In [20]:
# Which columns are categorical?
cate_cols=los3.select_dtypes("object").columns.tolist()
cate_cols

['customer',
 'state',
 'gender',
 'education',
 'customer_lifetime_value',
 'number_of_open_complaints',
 'policy_type',
 'vehicle_class']

In [21]:
# Understand the meaning of all columns

#customer: ID del cliente
#state: estado (dentro de Estados Unidos) de residencia del cliente
#gender: género del cliente. f para female y m para male (solo aparecen esos 2)
#education: nivel educativo del cliente
#customer_lifetime_value: valor total que aporta el cliente a la empresa
#income: ingreso generado por el cliente
#monthly_premium_auto: es la cantidad que paga el cliente a un proveedor de seguros 
#number_of_open_complaints: cantidad actual de quejas abiertos del cliente con la empresa
#policy_type: tipo de póliza de seguro
#vehicle_class: clase del vehículo del cliente
#total claim amount: el monto total reclamado por el cliente para la cobertura del seguro

Perform the data cleaning operations mentioned so far in class

In [22]:
#Delete the column education and the number of open complaints from the dataframe.
los3=los3.drop(["education", "number_of_open_complaints"], axis=1)
los3.columns

Index(['customer', 'state', 'gender', 'customer_lifetime_value', 'income',
       'monthly_premium_auto', 'policy_type', 'vehicle_class',
       'total_claim_amount'],
      dtype='object')

In [23]:
#Con ayuda de Sandra
def clean_customer_lifetime_value(value):
    if isinstance(value,str):
        #Remove percentage sign
        value = value.replace("%","")
        #Transform str to float
        value = float(value)
    value = value*100
    return value
los3["customer_lifetime_value"] = los3["customer_lifetime_value"].apply(clean_customer_lifetime_value)
los3["customer_lifetime_value"]

0                NaN
1       6.979536e+07
2       1.288743e+08
3       7.645862e+07
4       5.363076e+07
            ...     
7065    2.340599e+06
7066    3.096511e+05
7067    8.163890e+05
7068    7.524442e+05
7069    2.611837e+05
Name: customer_lifetime_value, Length: 12074, dtype: float64

In [24]:
print(los3["customer_lifetime_value"].dtypes)

float64


Ahora dio float64.

In [25]:
los3.isna().sum()

customer                   2937
state                      2937
gender                     3059
customer_lifetime_value    2944
income                     2937
monthly_premium_auto       2937
policy_type                2937
vehicle_class              2937
total_claim_amount         2937
dtype: int64

In [26]:
#Check for duplicate rows in the data and remove if any.
los3=los3.drop_duplicates()
print(los3.shape[0])

9135


Hubieron duplicados pues shape[0] disminuyó.

In [27]:
#Filter out the data for customers who have an income of 0 or less.
filtrados=los3[los3["income"]<=0]
filtrados

customer       state gender  customer_lifetime_value  income  \
0     RB50392  Washington    NaN                      NaN     0.0   
1     QZ44356     Arizona      F             6.979536e+07     0.0   
3     WW63253  California      M             7.645862e+07     0.0   
7     CF85061     Arizona      M             7.216100e+07     0.0   
10    SX51350  California      M             4.738992e+07     0.0   
...       ...         ...    ...                      ...     ...   
7059  WZ45103  California      F             5.678050e+05     0.0   
7061  RX91025  California      M             1.987226e+06     0.0   
7062  AC13887  California      M             4.628995e+05     0.0   
7067  TD14365  California      M             8.163890e+05     0.0   
7069  Y167826  California      M             2.611837e+05     0.0   

      monthly_premium_auto     policy_type  vehicle_class  total_claim_amount  
0                   1000.0   Personal Auto  Four-Door Car            2.704934  
1                     94.0   Personal Auto  Four-Door Car         1131.464935  
3                    106.0  Corporate Auto            SUV          529.881344  
7                    101.0  Corporate Auto  Four-Door Car          363.029680  
10                    67.0   Personal Auto  Four-Door Car          482.400000  
...                    ...             ...            ...                 ...  
7059                  76.0   Personal Auto  Four-Door Car          364.800000  
7061                 185.0   Personal Auto            SUV         1950.725547  
7062                  67.0  Corporate Auto   Two-Door Car          482.400000  
7067                  85.0  Corporate Auto  Four-Door Car          790.784983  
7069                  77.0  Corporate Auto   Two-Door Car          369.600000  

[2294 rows x 9 columns]

In [28]:
#Veamos si hay negativos (puestos por error)
negativos= los3[los3["income"]<0]
negativos.shape[0]

0

No hay, por lo que hay 2294 observaciones con Income=0.

Voy a corregir los valores de "gender" para que queden como "f" o "m" (poseriormente se tratarán los NaNs)

In [30]:
los3["gender"].unique()

array([nan, 'F', 'M', 'Femal', 'Male', 'female'], dtype=object)

In [31]:
los3["gender"]=los3["gender"].str.lower() 
los3["gender"].replace({"female": "f","femal":"f", "male":'m'}, inplace=True)

In [32]:
los3["gender"].unique()

array([nan, 'f', 'm'], dtype=object)

In [33]:
#Check and deal with NaN values
los3.isna().sum()

customer                     1
state                        1
gender                     123
customer_lifetime_value      8
income                       1
monthly_premium_auto         1
policy_type                  1
vehicle_class                1
total_claim_amount           1
dtype: int64

In [34]:
nulos_en_porcentaje=(los3.isna().sum()/len(los3))*100
print(nulos_en_porcentaje)

customer                   0.010947
state                      0.010947
gender                     1.346470
customer_lifetime_value    0.087575
income                     0.010947
monthly_premium_auto       0.010947
policy_type                0.010947
vehicle_class              0.010947
total_claim_amount         0.010947
dtype: float64


La columna que tiene mayor porcentaje de nulos tiene ~1.35%, el cual es un valor pequeño. Se procede a eliminar todas las observaciones con nulos del DataFrame (no es necesario que se imputen valores ni se eliminen columnas porque los porcentajes son muy chicos).

In [35]:
los3.dropna(inplace=True)

In [36]:
los3.isna().sum()

customer                   0
state                      0
gender                     0
customer_lifetime_value    0
income                     0
monthly_premium_auto       0
policy_type                0
vehicle_class              0
total_claim_amount         0
dtype: int64

In [37]:
los3.shape[0]

9007

Datetime format - Extract the months from the dataset marketing_customer_analysis.csv and store in a separate column. Then filter the data to show only the information for the first quarter , ie. January, February and March. _Hint_: If data from March does not exist, consider only January and February.

In [38]:
mktg=pd.read_csv("Data/marketing_customer_analysis.csv")
mktg.sample()

Unnamed: 0 Customer   State  Customer Lifetime Value Response Coverage  \
7479        7479  KF80172  Nevada              2647.519741      Yes    Basic   

     Education Effective To Date EmploymentStatus Gender  ...  \
7479   College           2/20/11         Employed      M  ...   

      Number of Open Complaints Number of Policies    Policy Type  \
7479                        0.0                  1  Personal Auto   

           Policy  Renew Offer Type  Sales Channel  Total Claim Amount  \
7479  Personal L3            Offer2          Agent          300.056383   

      Vehicle Class Vehicle Size Vehicle Type  
7479  Four-Door Car      Medsize          NaN  

[1 rows x 26 columns]

In [39]:
mktg.columns

Index(['Unnamed: 0', 'Customer', 'State', 'Customer Lifetime Value',
       'Response', 'Coverage', 'Education', 'Effective To Date',
       'EmploymentStatus', 'Gender', 'Income', 'Location Code',
       'Marital Status', 'Monthly Premium Auto', 'Months Since Last Claim',
       'Months Since Policy Inception', 'Number of Open Complaints',
       'Number of Policies', 'Policy Type', 'Policy', 'Renew Offer Type',
       'Sales Channel', 'Total Claim Amount', 'Vehicle Class', 'Vehicle Size',
       'Vehicle Type'],
      dtype='object')

In [40]:
mktg["month"] = pd.DatetimeIndex(mktg["Effective To Date"]).month
mktg.sample()

Unnamed: 0 Customer   State  Customer Lifetime Value Response Coverage  \
9153        9153  FX26050  Oregon              22712.02364       No    Basic   

     Education Effective To Date EmploymentStatus Gender  ...  \
9153    Doctor           1/16/11         Disabled      M  ...   

      Number of Policies   Policy Type      Policy  Renew Offer Type  \
9153                   2  Special Auto  Special L1            Offer4   

      Sales Channel  Total Claim Amount  Vehicle Class  Vehicle Size  \
9153          Agent           68.019973   Two-Door Car       Medsize   

     Vehicle Type month  
9153          NaN     1  

[1 rows x 27 columns]

In [41]:
mktg.month.unique()
#En el output solo aparecen los meses 1 y 2, es decir Enero y Febrero.

array([2, 1])

In [42]:
mktg[(mktg["month"]==1) | (mktg["month"]==2)]

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       Number of Policies     Policy Type        Policy  Renew Offer Type  \
0                       9  Corporate Auto  Corporate L3            Offer3   
1                       1   Personal Auto   Personal L3            Offer4   
2                       2   Personal Auto   Personal L3            Offer3   
3                       2  Corporate Auto  Corporate L3            Offer2   
4                       7   Personal Auto   Personal L2            Offer1   
...                   ...             ...           ...               ...   
10905                   7   Personal Auto   Personal L1            Offer3   
10906                   6   Personal Auto   Personal L3            Offer2   
10907                   2  Corporate Auto  Corporate L3            Offer1   
10908                   6   Personal Auto   Personal L1            Offer1   
10909                   3   Personal Auto   Personal L1            Offer4   

       Sales Channel  Total Claim Amount  Vehicle Class  Vehicle Size  \
0              Agent          292.800000  Four-Door Car       Medsize   
1        Call Center          744.924331  Four-Door Car       Medsize   
2        Call Center          480.000000            SUV       Medsize   
3             Branch          484.013411  Four-Door Car       Medsize   
4             Branch          707.925645  Four-Door Car       Medsize   
...              ...                 ...            ...           ...   
10905            Web         1214.400000     Luxury Car       Medsize   
10906         Branch          273.018929  Four-Door Car       Medsize   
10907            Web          381.306996     Luxury SUV       Medsize   
10908         Branch          618.288849            SUV       Medsize   
10909            Web         1021.719397            SUV       Medsize   

      Vehicle Type month  
0              NaN     2  
1              NaN     1  
2                A     2  
3                A     1  
4              NaN     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907          NaN     2  
10908            A     2  
10909          NaN     1  

[10910 rows x 27 columns]

In [43]:
print(mktg["Effective To Date"].dtypes)

object


In [44]:
mktg["Effective To Date"][5]

'2/14/11'

BONUS: Put all the previously mentioned data transformations into one or more functions.

In [45]:
#Standardize heather names
def titulos_cols(df):
    df.columns = df.columns.str.lower().str.replace(" ", "_")
    return df

In [46]:
#Rearrange columns in each DataFrame
def reordenar_cols(df, df_fijo):
    return df[df_fijo.columns]

In [47]:
#Esta se utilizó anteriormente
def clean_customer_lifetime_value(value):
    if isinstance(value,str):
        #Remove percentage sign
        value = value.replace("%","")
        #Transform str to float
        value = float(value)
    value = value*100
    return value
los3["customer_lifetime_value"] = los3["customer_lifetime_value"].apply(clean_customer_lifetime_value)
los3["customer_lifetime_value"]

1       6.979536e+09
2       1.288743e+10
3       7.645862e+09
4       5.363076e+09
5       8.256298e+09
            ...     
7065    2.340599e+08
7066    3.096511e+07
7067    8.163890e+07
7068    7.524442e+07
7069    2.611837e+07
Name: customer_lifetime_value, Length: 9007, dtype: float64

In [48]:
#Eliminar duplicados
def eliminar_duplicados(df):
    return df.drop_duplicates()

In [49]:
#Eliminar nulos
def eliminar_nulos(df):
    return df.dropna(inplace=True)